x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/respository/RemoteClusterRepository.java

```diff
@@ -0,0 +1,232 @@
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class RemoteClusterRepository extends AbstractLifecycleComponent implements Repository {
+
+    public static final String SNAPSHOT_UUID = "_latest_";
+    public static final String TYPE = "_remote_cluster_";
+
+    private final RepositoryMetaData metadata;
+    private final String remoteClusterAlias;
+    private final Client remoteClient;
+    private final CcrLicenseChecker ccrLicenseChecker;
+
+    public RemoteClusterRepository(RepositoryMetaData metadata, Client client, CcrLicenseChecker ccrLicenseChecker, Settings settings) {
+        super(settings);
+        this.metadata = metadata;
+        this.remoteClusterAlias = metadata.name();
+        this.ccrLicenseChecker = ccrLicenseChecker;
+        this.remoteClient = client.getRemoteClusterClient(remoteClusterAlias);
+    }
+
+    @Override
+    protected void doStart() {
+
+    }
+
+    @Override
+    protected void doStop() {
+
+    }
+
+    @Override
+    protected void doClose() throws IOException {
+
+    }
+
+    @Override
+    public RepositoryMetaData getMetadata() {
+        return metadata;
+    }
+
+    @Override
+    public SnapshotInfo getSnapshotInfo(SnapshotId snapshotId) {
+        assert SNAPSHOT_UUID.equals(snapshotId.getUUID()) : "RemoteClusterRepository only supports the _latest_ as the UUID";
+        // TODO: Perhaps add version
+        return new SnapshotInfo(snapshotId, Collections.singletonList(snapshotId.getName()), SnapshotState.SUCCESS);
+    }
+
+    @Override
+    public MetaData getSnapshotGlobalMetaData(SnapshotId snapshotId) {
+        assert SNAPSHOT_UUID.equals(snapshotId.getUUID()) : "RemoteClusterRepository only supports the _latest_ as the UUID";
+        ClusterStateResponse response = remoteClient.admin().cluster().prepareState().clear().setMetaData(true).get();
+        return response.getState().metaData();
+    }
+
+    @Override
+    public IndexMetaData getSnapshotIndexMetaData(SnapshotId snapshotId, IndexId index) {
+        assert SNAPSHOT_UUID.equals(snapshotId.getUUID()) : "RemoteClusterRepository only supports the _latest_ as the UUID";
+        String leaderIndex = index.getName();
+
+        ClusterStateResponse response = remoteClient
+            .admin()
+            .cluster()
+            .prepareState()
+            .clear()
+            .setMetaData(true)
+            .setIndices(leaderIndex)
+            .get();
+
+        // Validates whether the leader cluster has been configured properly:
+        PlainActionFuture<String[]> future = PlainActionFuture.newFuture();
+        IndexMetaData leaderIndexMetaData = response.getState().metaData().index(leaderIndex);
+        ccrLicenseChecker.fetchLeaderHistoryUUIDs(remoteClient, leaderIndexMetaData, future::onFailure, future::onResponse);
+        String[] leaderHistoryUuids = future.actionGet();
```

Comment:
is the fact that this is blocking here an issue?

Reply:
The snapshot/restore functionality appears to me to not be really designed to be asynchronous at this point. As the various methods do not support listeners. So I think the calls need to be blocking.

At this point we are on a `ThreadPool.Names.SNAPSHOT` thread which is a scaling thread pool that allows for 1 -> `Math.min(2 * num of cores, 5)`. So I think that this blocking is okay. Additionally all the other snapshot restore repositories appears to be stream/blocking based.

Obviously if we allow > 1 restore at a time, at some point a max of 5 threads + blocking might be a scaling problem. But I think improving that is beyond the scope of this PR.

Reply:
Got it, thanks for explaining!